builder/dockerfile/builder.go

```diff
@@ -0,0 +1,291 @@
+	DockerfileName string
+
+	Verbose     bool
+	UseCache    bool
+	Remove      bool
+	ForceRemove bool
+	Pull        bool
+	BuildArgs   map[string]string // build-time args received in build context for expansion/substitution and commands in 'run'.
+
+	// resource constraints
+	// TODO: factor out to be reused with Run ?
+
+	Memory       int64
+	MemorySwap   int64
+	CPUShares    int64
+	CPUPeriod    int64
+	CPUQuota     int64
+	CPUSetCpus   string
+	CPUSetMems   string
+	CgroupParent string
+	Ulimits      []*ulimit.Ulimit
+}
+
+// Builder is a Dockerfile builder
+// It implements the builder.Builder interface.
+type Builder struct {
+	*Config
+
+	Stdout io.Writer
+	Stderr io.Writer
+
+	docker  builder.Docker
+	context builder.Context
+
+	dockerfile       *parser.Node
+	runConfig        *runconfig.Config // runconfig for cmd, run, entrypoint etc.
+	flags            *BFlags
+	tmpContainers    map[string]struct{}
+	image            string // imageID
+	noBaseImage      bool
+	maintainer       string
+	cmdSet           bool
+	disableCommit    bool
+	cacheBusted      bool
+	cancelled        chan struct{}
+	cancelOnce       sync.Once
+	allowedBuildArgs map[string]bool // list of build-time args that are allowed for expansion/substitution and passing to commands in 'run'.
+
+	// TODO: remove once docker.Commit can receive a tag
+	id           string
+	activeImages []string
+}
+
+// NewBuilder creates a new Dockerfile builder from an optional dockerfile and a Config.
+// If dockerfile is nil, the Dockerfile specified by Config.DockerfileName,
+// will be read from the Context passed to Build().
+func NewBuilder(dockerfile io.ReadCloser, config *Config) (b *Builder, err error) {
+	if config == nil {
+		config = new(Config)
+	}
+	if config.BuildArgs == nil {
+		config.BuildArgs = make(map[string]string)
+	}
+	b = &Builder{
+		Config:           config,
+		Stdout:           os.Stdout,
+		Stderr:           os.Stderr,
+		runConfig:        new(runconfig.Config),
+		tmpContainers:    map[string]struct{}{},
+		cancelled:        make(chan struct{}),
+		id:               stringid.GenerateNonCryptoID(),
+		allowedBuildArgs: make(map[string]bool),
+	}
+	if dockerfile != nil {
+		b.dockerfile, err = parser.Parse(dockerfile)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return b, nil
+}
+
+// Build runs the Dockerfile builder from a context and a docker object that allows to make calls
+// to Docker.
+//
+// This will (barring errors):
+//
+// * read the dockerfile from context
+// * parse the dockerfile
+// * walk the parse tree and execute it by dispatching to handlers. If Remove
+//   or ForceRemove is set, additional cleanup around containers happens after
+//   processing.
+// * Print a happy message and return the image ID.
+// * NOT tag the image, that is responsibility of the caller.
+//
+func (b *Builder) Build(docker builder.Docker, context builder.Context) (builder.ImageID, error) {
+	b.docker = docker
```

Comment:
this looks weird as for me, why not setup them in `NewBuilder`? We create it every time anyway.